File: 445. Add Two Numbers II/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        rfirst, rsecond = self.reverse(l1), self.reverse(l2)
        cfirst, csecond = rfirst, rsecond
        pfirst, psecond = None, None
        carry = 0
        while cfirst and csecond:
            val = cfirst.val + csecond.val + carry
            carry = 1 if val > 9 else 0
            cfirst.val = val % 10
            pfirst, psecond = cfirst, csecond
            cfirst, csecond = cfirst.next, csecond.next
        if not cfirst and csecond:
            pfirst.next = csecond
            cfirst = csecond
        while cfirst or carry:
            if not cfirst:
                cfirst = ListNode()
                pfirst.next = cfirst
            val = cfirst.val + carry
            carry = 1 if val > 9 else 0
            cfirst.val = val % 10
            pfirst = cfirst
            cfirst = cfirst.next
        return self.reverse(rfirst)

    def reverse(self, root):
        prev, curr, foll = None, root, root
        while curr:
            foll = foll.next
            curr.next = prev
            prev = curr
            curr = foll
        return prev

File: 445. Add Two Numbers II/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode rfirst = reverse(l1), rsecond = reverse(l2);
        ListNode pfirst = null, psecond = null;
        ListNode cfirst = rfirst, csecond = rsecond;
        int carry = 0, val = 0;
        while (cfirst != null && csecond != null) {
            val = cfirst.val + csecond.val + carry;
            carry = (val > 9) ? 1 : 0;
            cfirst.val = val % 10;
            pfirst = cfirst;
            psecond = csecond;
            cfirst = cfirst.next;
            csecond = csecond.next;
        }
        if (cfirst == null && csecond != null) {
            pfirst.next = csecond;
            cfirst = csecond;
        }
        while (cfirst != null || carry > 0) {
            if (cfirst == null) {
                pfirst.next = new ListNode();
                cfirst = pfirst.next;
            }
            val = cfirst.val + carry;
            carry = (val > 9) ? 1 : 0;
            cfirst.val = val % 10;
            pfirst = cfirst;
            cfirst = cfirst.next;
        }
        return reverse(rfirst);
    }

    public ListNode reverse(ListNode root) {
        ListNode prev = null, curr = root, foll = root;
        while (curr != null) {
            foll = foll.next;
            curr.next = prev;
            prev = curr;
            curr = foll;
        }
        return prev;
    }
}
